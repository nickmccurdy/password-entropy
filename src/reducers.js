import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, { callback = identity, validator } = {}) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = !validator || validator(payload)

    return matchesAction && isValid ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6, { validator: n => n >= 1 }),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM', 1, { validator: n => n >= 0 }),
    generic: createReducer ('TOGGLE_GENERIC', map(T, GenericOptions.toggles), {
      callback: ({ checked, name }, state) => assoc(name, checked, state)
    })
  }),
  optionsKey: createReducer('SET_OPTIONS_KEY', 'diceware')
})
