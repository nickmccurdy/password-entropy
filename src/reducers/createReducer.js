import { identity } from 'ramda'

export default function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = typeof payload !== 'number' || payload >= 0

    return matchesAction && isValid ? callback(payload, state) : state
  }
}
