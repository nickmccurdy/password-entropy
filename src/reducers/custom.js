import { path } from 'ramda'

import createReducer from './createReducer'

export default createReducer('SET_CUSTOM', 0)

export function setCustom({ target: { value } }) {
  return { type: 'SET_CUSTOM', payload: parseInt(value, 10) }
}

export const selector = state => ({
  possibleItems: path(['options', 'custom'], state)
})
