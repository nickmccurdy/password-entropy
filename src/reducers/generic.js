import assoc from 'ramda/src/assoc'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'

import createReducer from './createReducer'

export const toggles = {
  Letters: {
    example: 'a-z',
    possibleItems: 26
  },
  'Capital Letters': {
    example: 'A-Z',
    possibleItems: 26
  },
  Numbers: {
    example: '0-9',
    possibleItems: 10
  },
  Symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function toggleGenericCallback({ checked, name }, state) {
  return assoc(name, checked, state)
}

export default createReducer(
  'TOGGLE_GENERIC',
  defaultToggles,
  toggleGenericCallback
)

export function toggleGeneric({ target: { name } }, checked) {
  return {
    type: 'TOGGLE_GENERIC',
    payload: {
      checked,
      name
    }
  }
}

export const selector = path(['options', 'Generic'])
