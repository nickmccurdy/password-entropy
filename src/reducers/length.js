import pick from 'ramda/src/pick'

import createReducer from './createReducer'

export default createReducer('SET_LENGTH', 12)

export function setLength({ target: { value } }) {
  return {
    type: 'SET_LENGTH',
    payload: parseInt(value, 10)
  }
}

export const selector = pick(['length'])
