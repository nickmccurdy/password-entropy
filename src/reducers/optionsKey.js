import pick from 'ramda/src/pick'

import createReducer from './createReducer'

export default createReducer('SET_OPTIONS_KEY', 'Generic')

export function setOptionsKey(event, key) {
  return {
    type: 'SET_OPTIONS_KEY',
    payload: key
  }
}

export const selector = pick(['optionsKey'])
