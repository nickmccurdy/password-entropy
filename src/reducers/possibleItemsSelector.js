import blue from 'material-ui/colors/blue'
import green from 'material-ui/colors/green'
import red from 'material-ui/colors/red'
import yellow from 'material-ui/colors/yellow'
import always from 'ramda/src/always'
import filter from 'ramda/src/filter'
import findLast from 'ramda/src/findLast'
import identity from 'ramda/src/identity'
import keys from 'ramda/src/keys'
import map from 'ramda/src/map'
import pipe from 'ramda/src/pipe'
import sum from 'ramda/src/sum'

import { toggles } from './generic'

export const dicewareWords = 7776

const entropyTips = [
  {
    color: red,
    minimum: 0,
    strength: 'Very Weak'
  },
  {
    color: yellow,
    minimum: 32,
    strength: 'Weak'
  },
  {
    color: blue,
    minimum: 64,
    strength: 'Strong'
  },
  {
    color: green,
    minimum: 128,
    strength: 'Very Strong'
  }
]

const possibleItems = {
  generic: pipe(
    filter(identity),
    keys,
    map(key => toggles[key].possibleItems),
    sum
  ),
  diceware: always(dicewareWords),
  custom: identity
}

export default function possibleItemsSelector({ length, options, optionsKey }) {
  const possiblePasswords = possibleItems[optionsKey](options[optionsKey]) ** length
  const entropyBits = Math.max(0, Math.log2(possiblePasswords))

  return {
    approximate: possiblePasswords > Number.MAX_SAFE_INTEGER,
    entropyBits,
    entropyTip: findLast(tip => entropyBits >= tip.minimum, entropyTips),
    possiblePasswords
  }
}
