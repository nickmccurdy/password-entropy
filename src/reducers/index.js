import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
