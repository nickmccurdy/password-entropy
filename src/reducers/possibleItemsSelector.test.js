import { mergeDeepRight } from 'ramda'
import { createStore } from 'redux'

import selector from './possibleItemsSelector'
import reducers from '.'

describe('selector', () => {
  function mergedSelector(state = {}) {
    return selector(mergeDeepRight(createStore(reducers).getState(), state))
  }

  test('generic', () => {
    expect(mergedSelector()).toMatchSnapshot()
  })

  test('diceware', () => {
    expect(mergedSelector({ optionsKey: 'diceware' })).toMatchSnapshot()
  })

  test('custom', () => {
    expect(mergedSelector({ optionsKey: 'custom' })).toMatchSnapshot()
  })

  test('custom with invalid possibleItems', () => {
    expect(
      mergedSelector({ options: { custom: 0 }, optionsKey: 'custom' })
    ).toMatchSnapshot()
  })

  test('custom with weak possiblePasswords', () => {
    expect(
      mergedSelector({
        length: 32,
        options: { custom: 2 },
        optionsKey: 'custom'
      })
    ).toMatchSnapshot()
  })

  test('custom with strong possiblePasswords', () => {
    expect(
      mergedSelector({
        length: 64,
        options: { custom: 2 },
        optionsKey: 'custom'
      })
    ).toMatchSnapshot()
  })

  test('custom with very strong possiblePasswords', () => {
    expect(
      mergedSelector({
        length: 128,
        options: { custom: 2 },
        optionsKey: 'custom'
      })
    ).toMatchSnapshot()
  })
})
