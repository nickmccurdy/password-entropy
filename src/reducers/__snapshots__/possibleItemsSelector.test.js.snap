// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selector custom 1`] = `
Object {
  "approximate": false,
  "entropyBits": 0,
  "entropyTip": Object {
    "color": "#f44336",
    "minimum": 0,
    "strength": "Very Weak",
  },
  "possiblePasswords": 0,
}
`;

exports[`selector custom with invalid possibleItems 1`] = `
Object {
  "approximate": false,
  "entropyBits": 0,
  "entropyTip": Object {
    "color": "#f44336",
    "minimum": 0,
    "strength": "Very Weak",
  },
  "possiblePasswords": 0,
}
`;

exports[`selector custom with strong possiblePasswords 1`] = `
Object {
  "approximate": true,
  "entropyBits": 64,
  "entropyTip": Object {
    "color": "#2196f3",
    "minimum": 64,
    "strength": "Strong",
  },
  "possiblePasswords": 18446744073709552000,
}
`;

exports[`selector custom with very strong possiblePasswords 1`] = `
Object {
  "approximate": true,
  "entropyBits": 128,
  "entropyTip": Object {
    "color": "#4caf50",
    "minimum": 128,
    "strength": "Very Strong",
  },
  "possiblePasswords": 3.402823669209385e+38,
}
`;

exports[`selector custom with weak possiblePasswords 1`] = `
Object {
  "approximate": false,
  "entropyBits": 32,
  "entropyTip": Object {
    "color": "#ffeb3b",
    "minimum": 32,
    "strength": "Weak",
  },
  "possiblePasswords": 4294967296,
}
`;

exports[`selector diceware 1`] = `
Object {
  "approximate": true,
  "entropyBits": 155.09775004326937,
  "entropyTip": Object {
    "color": "#4caf50",
    "minimum": 128,
    "strength": "Very Strong",
  },
  "possiblePasswords": 4.887367798068926e+46,
}
`;

exports[`selector generic 1`] = `
Object {
  "approximate": true,
  "entropyBits": 73.55139620333959,
  "entropyTip": Object {
    "color": "#2196f3",
    "minimum": 64,
    "strength": "Strong",
  },
  "possiblePasswords": 1.3841287201e+22,
}
`;
